control-plane/api-gateway/controllers/gateway_class_config_controller.go

```diff
@@ -0,0 +1,139 @@
+		if err != nil {
+			r.Log.Error(err, "failed to check if the gateway class config is still in use")
+			return ctrl.Result{}, err
+		}
+		if used {
+			r.Log.Info("gateway class config still in use")
+			// Requeue as to not block the reconciliation loop.
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		// gcc is no longer in use.
+		if _, err := r.RemoveFinalizer(ctx, gcc, gatewayClassConfigFinalizer); err != nil {
+			r.Log.Error(err, "error removing gateway class config finalizer")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if _, err := r.EnsureFinalizer(ctx, gcc, gatewayClassConfigFinalizer); err != nil {
+		r.Log.Error(err, "error adding gateway class config finalizer")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Controller) EnsureFinalizer(ctx context.Context, object client.Object, finalizer string) (bool, error) {
+	finalizers := object.GetFinalizers()
+	for _, f := range finalizers {
+		if f == finalizer {
+			return false, nil
+		}
+	}
+	object.SetFinalizers(append(finalizers, finalizer))
+	if err := r.Update(ctx, object); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// RemoveFinalizer ensures that the given finalizer is removed from the passed object
+// it returns a boolean saying whether a finalizer was removed, and any
+// potential errors.
+func (r *Controller) RemoveFinalizer(ctx context.Context, object client.Object, finalizer string) (bool, error) {
+	finalizers := []string{}
+	found := false
+	for _, f := range object.GetFinalizers() {
+		if f == finalizer {
+			found = true
+			continue
+		}
+		finalizers = append(finalizers, f)
+	}
+	if found {
+		object.SetFinalizers(finalizers)
+		if err := r.Update(ctx, object); err != nil {
+			return false, err
+		}
+	}
+	return found, nil
+}
+
+// gatewayClassUsesConfig determines whether a given GatewayClass references a
+// given GatewayClassConfig. Since these resources are scoped to the cluster,
+// namespace is not considered.
+func gatewayClassUsesConfig(gc gwv1beta1.GatewayClass, gcc *v1alpha1.GatewayClassConfig) bool {
+	paramaterRef := gc.Spec.ParametersRef
+	return paramaterRef != nil &&
+		string(paramaterRef.Group) == v1alpha1.ConsulHashicorpGroup &&
+		paramaterRef.Kind == v1alpha1.GatewayClassConfigKind &&
+		paramaterRef.Name == gcc.Name
+}
+
+// GatewayClassConfigInUse determines whether any GatewayClass in the cluster
+// references the provided GatewayClassConfig.
+func (r *Controller) GatewayClassConfigInUse(ctx context.Context, gcc *v1alpha1.GatewayClassConfig) (bool, error) {
+	list := &gwv1beta1.GatewayClassList{}
+	if err := r.List(ctx, list); err != nil {
+		return false, err
+	}
+
+	for _, gc := range list.Items {
+		if gatewayClassUsesConfig(gc, gcc) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (r *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.GatewayClassConfig{}).
```

Comment:
We also want to cover `v1beta2.GatewayClassConfig{}` and maybe `v1beta1.GatewayClassConfig{}`.

Reply:
So, this is our internal CRD, not anything from the spec and all of our internal CRDs are `v1alpha1`, so this should be fine.

Reply:
Oh gosh, that's right. I was just thinking of this in my own GatewayClass Controller code and got mixed up.